pkg/autoupdate/tuf_autoupdate.go

```diff
@@ -0,0 +1,211 @@
+	checkInterval   time.Duration
+	errorCounter    []int64
+	lock            sync.RWMutex
+	interrupt       chan struct{}
+	logger          log.Logger
+}
+
+type TufAutoupdaterOption func(*TufAutoupdater)
+
+func WithTufLogger(logger log.Logger) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.logger = log.With(logger, "component", "tuf_autoupdater")
+	}
+}
+
+func WithChannel(channel UpdateChannel) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.channel = channel
+	}
+}
+
+func WithUpdateCheckInterval(checkInterval time.Duration) TufAutoupdaterOption {
+	return func(ta *TufAutoupdater) {
+		ta.checkInterval = checkInterval
+	}
+}
+
+func NewTufAutoupdater(metadataUrl, mirrorUrl, binaryPath, rootDirectory string, opts ...TufAutoupdaterOption) (*TufAutoupdater, error) {
+	binaryName := filepath.Base(binaryPath)
+
+	// Set up the local TUF directory for our TUF client -- a dev repo, to be replaced once we move to production
+	localTufDirectory := filepath.Join(rootDirectory, fmt.Sprintf("%s-tuf-dev", strings.TrimSuffix(binaryName, ".exe")))
+	if err := os.MkdirAll(localTufDirectory, 0750); err != nil {
+		return nil, fmt.Errorf("could not make local TUF directory %s: %w", localTufDirectory, err)
+	}
+
+	// Set up our local store i.e. point to the directory in our filesystem
+	localStore, err := filejsonstore.NewFileJSONStore(localTufDirectory)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize local TUF store: %w", err)
+	}
+
+	// Set up our remote store i.e. tuf-devel.kolide.com
+	remoteOpts := client.HTTPRemoteOptions{
+		MetadataPath: "/repository",
+	}
+	remoteStore, err := client.HTTPRemoteStore(metadataUrl, &remoteOpts, http.DefaultClient)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize remote TUF store: %w", err)
+	}
+
+	metadataClient := client.NewClient(localStore, remoteStore)
+	if err := metadataClient.Init(rootJson); err != nil {
+		return nil, fmt.Errorf("failed to initialize TUF client with root JSON: %w", err)
+	}
+
+	ta := &TufAutoupdater{
+		metadataClient:  metadataClient,
+		mirrorClient:    http.DefaultClient,
+		mirrorUrl:       mirrorUrl,
+		binary:          binaryName,
+		operatingSystem: runtime.GOOS,
+		channel:         Stable,
+		interrupt:       make(chan struct{}),
+		checkInterval:   60 * time.Second,
+		errorCounter:    make([]int64, 0), // For now, the error counter is a simple list of timestamps when errors occurred
+		logger:          log.NewNopLogger(),
+	}
+
+	for _, opt := range opts {
+		opt(ta)
+	}
+
+	return ta, nil
+}
+
+func (ta *TufAutoupdater) Run(opts ...legacytuf.Option) (stop func(), err error) {
+	go ta.loop()
+
+	return ta.stop, nil
+}
+
+func (ta *TufAutoupdater) RollingErrorCount() int {
+	ta.lock.RLock()
+	defer ta.lock.RUnlock()
+
+	oneDayAgo := time.Now().Add(-24 * time.Hour).Unix()
+	errorCount := 0
+	for _, errorTimestamp := range ta.errorCounter {
+		if errorTimestamp < oneDayAgo {
+			continue
+		}
+		errorCount += 1
+	}
+
```

Comment:
Does anything ever clean `errorCounter`?

Reply:
No, just a launcher or computer restart -- I wasn't super worried given the long default update interval and given that this will be updated in the next couple PRs (though that sort of comment will jinx it to hang around forever 😀). I can add some cleanup here.

Reply:
I want to say it will be okay, but I'm sure they'll be _one_ person who has intermittent network, and it causes an endless stream of errors and never restarts.

Reply:
Makes sense, better to be on the safe side! I've added a cleanup job that runs 2x a day